### PR TITLE
Default prompt of environment name in batch shell

### LINKED ISSUE #1570: No (env) indication visible - Windows 10
In all the time I've used virtualenv (i.e since 2 years now), whenever I activate a virtual environment there is an (env) indication like this:
image

But with the new virtualenv version (20.0.1), I have no indication of whether the virtualenv is active or not: 
image

Was this intended? How do I get the previous functionality back? virtualenv version 20 seems much different than 16.

### PR DESCRIPTION
**Pull request by spetafree**
Fixes #1570, by setting a default prompt for batch (windows cmd.exe) shell based on the environment name.

In bash activation script, the check for using the default prompt is done in the bash shell script. It is more complicated to check that and get the environment folder name in bat script, and because of that I have decided to implement that check in the python code, by inheriting the `replacements` method and adding the check there.

I had a few failures in the local tests I tried to run, but is seems to me that these are not connected to my change - the same tests happen also on the current master.

### Thanks for contributing, make sure you address all the checklists (for details on how see
[development documentation](https://virtualenv.pypa.io/en/latest/development.html#development))!

- [X] ran the linter to address style issues (``tox -e fix_lint``)

- [X] wrote descriptive pull request text

Above.

- [ ] ensured there are test(s) validating the fix

I am actually at loss how to test this, the current tests seem not to test that issue also on other
shells where I assume it works.

- [X] added news fragment in ``docs/changelog`` folder
I do not understand how should I name the fragment, I hope it is ok.

- [X] updated/extended the documentation
n/a

### CODE REVIEW

File: src/virtualenv/activation/batch/__init__.py [hunk 1]
```diff
@@ -21,3 +21,9 @@ def instantiate_template(self, replacements, template, creator):
         # ensure the text has all newlines as \r\n - required by batch
         base = super(BatchActivator, self).instantiate_template(replacements, template, creator)
         return base.replace(os.linesep, "\n").replace("\n", os.linesep)
+
+    def replacements(self, creator, dest_folder):
+        base_replacements = super(BatchActivator, self).replacements(creator, dest_folder)
+        if not base_replacements["__VIRTUAL_PROMPT__"]:
+            base_replacements["__VIRTUAL_PROMPT__"] = "(%s) " % creator.env_name
```
Comment (gaborbernat) [inline src/virtualenv/activation/batch/__init__.py]: Shouldn't this change be in the template? Not the parameterization?

Comment (spetafree) [reply src/virtualenv/activation/batch/__init__.py]: I am not sure what do you mean, I will try to explain my reasoning:

In all other shells, there is a check in the shell script if the value replaced by virtual prompt is empty, and if so, it is set to the `env_name` in bat it is hard to do it, at least to my limited expertise in bat programing. To overcome this I have made the check here.

I do not think this should be in the `via_template.py` file if that what you mean, I see no need to complicate other shells because of bat limitations.

Comment (spetafree) [reply src/virtualenv/activation/batch/__init__.py]: A possible different change would be to change
https://github.com/pypa/virtualenv/blob/7649968fcc673336e510a18d4d1d20feef48fe4b/src/virtualenv/activation/via_template.py#L34
```python
 "__VIRTUAL_PROMPT__": "" if self.flag_prompt is None else self.flag_prompt,
```
to

````python
 "__VIRTUAL_PROMPT__": '(%s) ' % creator.env if self.flag_prompt is None else self.flag_prompt,
```
But I am not sure and will be unable to check how it effects all other shells, and the only ones I have access to are cmd.exe and (git) bash (both on windows).

Comment (spetafree) [reply src/virtualenv/activation/batch/__init__.py]: > 
> 
> Was thinking personally this should be a change within activate.bat 🤔

Correctly checking that a variable is empty and getting the folder name in bat is rather ugly, but I can try doing it, I'll need to learn how to. Seemed to me cleaner that way, similar to what have been in the previous version of vritualenv - the windows bat case was handled in python.

Comment (gaborbernat) [reply src/virtualenv/activation/batch/__init__.py]: Leave it as it is then, will investigate next week 👍

**Comment by gaborbernat** [comment]
I'm not saying we absolutely must do it in batch but would like to see the two options 👍and how they would differ and if it's worth doing differently here than for all other scripts.

**Comment by spetafree** [comment]
What do you think about the change in `via_template.py`? Maybe it will fix all the shells at once? The prompt will just always default to '(env_name) ' instead of to empty string in python, and the shells scripts may be simplified.

**Comment by gaborbernat** [comment]
Can't tell from top of my head at the moment. Need to have a think about it.

**Comment by spetafree** [comment]
It may take me some time, but may try to install some of the other shells and see how it works. There is also the question of how to specify an empty prompt in command line. (maybe we will need a --no-prompt flag)
Thanks for reviewing the PR anyway, It will probably be a few days at least until I can learn enough about all these shells, I'm much more familiar with python that with shell scripts 😄

**Comment by gaborbernat** [comment]
No worries, take your time, I'll not be able to take a deeper look at this until next Tuesday.

File: src/virtualenv/activation/batch/__init__.py (hunk 1 above)
Comment (spetafree) [reply src/virtualenv/activation/batch/__init__.py]: Turns out fixing it it in the bat file, at least for the simple case was not so hard. Although, I can't say I understand why is this exact loop works and other I tried failed, so I'm not sure that it will work in edge cases, if there are.

I hope it's better now

**Comment by spetafree** [comment]
I'll try to write the test tomorrow. Do you think that adding another command to print the prompt environment variable in _get_test_lines and assert it in assert_output in `test_batch.py` is the correct way?

**Comment by spetafree** [comment]
Any chance you'll consider re-opening it, I was not able to work on it the last couple of days? Tomorrow I'll hopefully be able to work on it again.

**Comment by spetafree** [comment]
I'll explain the changes:

1. I had to change the handling of the variables in the batch script to support the non ascii characters. The following seems to work:

https://github.com/pypa/virtualenv/blob/dddf7cdacdb0042a19e9a946721f8898a349bd02/src/virtualenv/activation/batch/activate.bat#L17

2.  The folder name in tests:  `e-$ èрт🚒♞中片-j` contains a single `$` character which is an escape character in batch prompts, and causes incorrect rendering. I have changed it to `$$`. I hope it's ok with all other shells, and doesn't drop some edge case that was checked with single `$`.

3. I have added testing with and without `--prompt` to all shells.

4. I have expanded the test for batch activation test to check for correct prompt. I have not added the same test to other shells. I can do it for bash, which I am somewhat familiar with, but I do not know how to test this for other shells.

p.s. @gaborbernat,  This is my second ever attempt at a public open source pull request, and I would like to thank you very much for your handling of it 😄. It is a much nicer treatment than my last attempt...

**Comment by spetafree** [comment]
I see I missed something. Will fix in a few hours.

**Comment by spetafree** [comment]
I have no idea why to the zipapp test fails, and how is this related to my PR. 

Any help will be appreciated, I will continue investigating. But currently I can't reproduce this on my windows machine.

**Comment by spetafree** [comment]
@gaborbernat Thanks a lot 😄.

But, there is no test for the prompt now, why did you decide to approve? 

Also the parameteriztion in conftest.py change has no effect now as far as I understand.

**Comment by gaborbernat** [comment]
Why would it not have any effect, I think this is alright 🤔

**Comment by spetafree** [comment]
Because no test looks at the prompt now as far as I understand. 

The only test that actually checked the prompt was in this commit:

https://github.com/pypa/virtualenv/pull/1679/commits/dddf7cdacdb0042a19e9a946721f8898a349bd02

**Comment by gaborbernat** [comment]
That's not true, the tests already looked at it, just did not run the test suite Without prompt flag.

**Comment by spetafree** [comment]
Ok, then I couldn't understand the tests - I didn't find a test that actually reads the prompt. I'll look again. Thanks anyway.

**Comment by gaborbernat** [comment]
Hello, a fix for this issue has been released via virtualenv 20.0.8; see https://pypi.org/project/virtualenv/20.0.8/ (https://virtualenv.pypa.io/en/latest/changelog.html#v20-0-8-2020-03-04). Please give a try and report back if your issue has not been addressed; if not, please comment here, and we'll reopen the ticket. We want to apologize for the inconvenience this has caused you and say thanks for having patience while we resolve the unexpected bugs with this new major release.
thanks
